packages/server/server/src/server/devServer.ts

```diff
@@ -84,48 +115,65 @@ export class ModernDevServer extends ModernServer {
     });
   }
 
-  private applySetupMiddlewares() {
-    const setupMiddlewares = this.dev.setupMiddlewares || [];
+  // Complete the preparation of services
+  public async onInit(runner: ServerHookRunner, app: Server) {
+    this.runner = runner;
+    const { dev, conf } = this;
 
-    const serverOptions: ExposeServerApis = {
-      sockWrite: (type, data) => this.devMiddleware.sockWrite(type, data),
-    };
+    // the http-compression can't handler stream http.
+    // so we disable compress when user use stream ssr temporarily.
+    const isUseStreamingSSR = (routes?: ServerRoute[]) =>
+      routes?.some(r => r.isStream === true);
 
-    const befores: RequestHandler[] = [];
-    const afters: RequestHandler[] = [];
+    const isUseSSRPreload = () => {
+      const {
+        server: { ssr, ssrByEntries },
+      } = conf;
 
-    setupMiddlewares.forEach(handler => {
-      handler(
-        {
-          unshift: (...handlers) => befores.unshift(...handlers),
-          push: (...handlers) => afters.push(...handlers),
-        },
-        serverOptions,
-      );
-    });
+      const checkUsePreload = (ssr?: SSR) =>
+        typeof ssr === 'object' && Boolean(ssr.preload);
 
-    return { befores, afters };
-  }
+      return (
+        checkUsePreload(ssr) ||
+        Object.values(ssrByEntries || {}).some(ssr => checkUsePreload(ssr))
+      );
+    };
 
-  // Complete the preparation of services
-  public async onInit(runner: ServerHookRunner, app: Server) {
-    this.runner = runner;
+    const { middlewares: rsbuildMiddlewares, close, onUpgrade } =
+      // https://github.com/web-infra-dev/rsbuild/blob/32fbb85e22158d5c4655505ce75e3452ce22dbb1/packages/shared/src/types/server.ts#L112
+      await this.getMiddlewares({
+        ...transformToRsbuildServerOptions(this.dev),
+        compress:
+          !isUseStreamingSSR(this.getRoutes()) &&
+          !isUseSSRPreload() &&
+          dev.compress,
+        htmlFallback: false,
+        publicDir: false,
+      });
 
-    const { dev } = this;
+    app.on('upgrade', onUpgrade);
 
-    // Order: devServer.before => setupMiddlewares.unshift => internal middlewares => setupMiddlewares.push => devServer.after
-    const { befores, afters } = this.applySetupMiddlewares();
+    this.rsbuild.onDevCompileDone(({ stats }) => {
+      // Reset only when client compile done
+      if (stats.toJson({ all: false }).name !== 'server') {
+        this.onRepack({ routes: this.getRoutes() });
+      }
+    });
 
     // before dev handler
     const beforeHandlers = await this.setupBeforeDevMiddleware();
```

Comment:
It seems that the order here is inconsistent.
Origin: before - > hookBefore - > unshift - > push - > hookAfter - > after
Now: hookBefore - > before - > unshift - > push - > after - > hookAfter

However, these two hooks are not used. we can remove them, and developer won't need to use them later.

Reply:
emm... I will remove these two hooks  in next PR

Reply:
updated!